src/components/tooltip/tooltip.js

```diff
@@ -0,0 +1,198 @@
+      .isRequired,
+    leaveDelay: PropTypes.number.isRequired,
+    isOpen: PropTypes.bool,
+    onClose: PropTypes.func,
+    onOpen: PropTypes.func,
+    placement: PropTypes.oneOf([
+      'top',
+      'top-start',
+      'top-end',
+      'right',
+      'right-start',
+      'right-end',
+      'bottom',
+      'bottom-start',
+      'bottom-end',
+      'left',
+      'left-start',
+      'left-end',
+    ]),
+    title: PropTypes.string.isRequired,
+    components: PropTypes.shape({
+      WrapperComponent: (props, propName) => {
+        if (props[propName] && !isValidElementType(props[propName])) {
+          return new Error(
+            `Invalid prop 'components.WrapperComponent' supplied to 'WrapperComponent': the prop is not a valid React component`
+          );
+        }
+        return null;
+      },
+    }),
+  };
+
+  static defaultProps = {
+    components: {},
+    leaveDelay: 0,
+    horizontalConstraint: 'scale',
+    placement: 'top',
+    type: 'info',
+  };
+
+  static getDerivedStateFromProps = (props, state) => ({
+    id: getFieldId(props, state, sequentialId),
+  });
+
+  state = {
+    open: this.isControlled ? this.props.isOpen : false,
+  };
+
+  componentWillUnmount() {
+    clearTimeout(this.leaveTimer);
+  }
+
+  handleEnter = event => {
+    const childrenProps = this.props.children.props;
+
+    // Remove the title ahead of time.
+    // We don't want to wait for the next render commit.
+    // We would risk displaying two tooltips at the same time (native + this one).
+    if (this.childrenRef && typeof this.childrenRef === 'function') {
+      this.childrenRef.setAttribute('title', '');
+    }
+
+    if (event.type === 'mouseover' && childrenProps.onMouseOver) {
+      childrenProps.onMouseOver(event);
+    }
+
+    if (event.type === 'focus' && childrenProps.onFocus) {
+      childrenProps.onFocus(event);
+    }
+
+    if (!this.state.open && !this.isControlled) {
+      this.setState({
+        open: true,
+      });
+    }
+
+    if (this.props.onOpen) {
+      this.props.onOpen(event);
+    }
+
+    event.preventDefault();
+    event.stopPropagation();
+  };
+
+  handleLeave = event => {
+    const childrenProps = this.props.children.props;
+    clearTimeout(this.leaveTimer);
+
+    if (event.type === 'mouseleave' && childrenProps.onMouseLeave) {
+      childrenProps.onMouseLeave(event);
+    }
+
+    if (event.type === 'blur' && childrenProps.onBlur) {
+      childrenProps.onBlur(event);
+    }
+
+    if (this.props.leaveDelay) {
+      this.leaveTimer = setTimeout(() => {
+        this.handleClose();
+      }, this.props.leaveDelay);
+    } else {
+      this.handleClose();
+    }
+  };
+
+  handleClose = event => {
+    if (!this.isControlled) {
+      this.setState({
+        open: false,
+      });
+    }
+    if (this.props.onClose) {
+      this.props.onClose(event);
+    }
+  };
+
+  setChildrenRef = ref => {
+    this.childrenRef = ref;
+  };
+
+  isControlled = this.props.isOpen != null;
```

Comment:
```suggestion
  isControlled = Boolean(this.props.isOpen);
```

Reply:
@emmenko but that isn't what we want?

If the user passed in `isOpen`, then they have to control it's open and closed state. 

So if I pass in false, I want to control it myself.

```

this.props.isOpen != null true
tooltip.js:141 Boolean(this.props.isOpen)) false
```

Reply:
Oh nvm, now I see the `!=` loose comparator. 🤦‍♂️
All good, I was wondering why checking only for `null` and e.g. not for `undefined`.